### PR TITLE
Search for target_triple.json only if builtin target not found

### PR DESCRIPTION
**Pull request by gnzlbg**
Before this commit, if the builtin target was found, but an error
happened when instantiating it (e.g. validating the target
specification file failed, etc.), then we ignored those errors
and proceeded to try to find a `target_triple.json` file, and if
that failed, reported that as an error.

With this commit, if rustc is supposed to provide the builtin target,
and something fails while instantiating it, that error will
get properly propagated.

r? @oli-obk

### CODE REVIEW

**Comment by gnzlbg** [comment]
cc @alexcrichton

File: src/librustc_target/spec/mod.rs [hunk 1]
```diff
@@ -269,21 +274,24 @@ macro_rules! supported_targets {
...
         const TARGETS: &[&str] = &[$($triple),*];
 
-        fn load_specific(target: &str) -> TargetResult {
+        fn load_specific(target: &str) -> Result<Target, LoadTargetError> {
             match target {
                 $(
                     $triple => {
-                        let mut t = $module::target()?;
+                        let mut t = $module::target()
+                            .map_err(|e| LoadTargetError::Other(e))?;
```
Comment (oli-obk) [inline src/librustc_target/spec/mod.rs]:
```suggestion
                            .map_err(LoadTargetError::Other)?;
```

File: src/librustc_target/spec/mod.rs [hunk 2]
```diff
@@ -269,21 +274,24 @@ macro_rules! supported_targets {
...
                         t.options.is_builtin = true;
 
                         // round-trip through the JSON parser to ensure at
                         // run-time that the parser works correctly
-                        t = Target::from_json(t.to_json())?;
+                        t = Target::from_json(t.to_json())
+                            .map_err(|e| LoadTargetError::Other(e))?;
```
Comment (oli-obk) [inline src/librustc_target/spec/mod.rs]:
```suggestion
                            .map_err(LoadTargetError::Other)?;
```

File: src/librustc_target/spec/mod.rs [hunk 3]
```diff
@@ -1176,8 +1184,10 @@ impl Target {
         match *target_triple {
             TargetTriple::TargetTriple(ref target_triple) => {
                 // check if triple is in list of supported targets
-                if let Ok(t) = load_specific(target_triple) {
-                    return Ok(t)
+                match load_specific(target_triple) {
+                    Ok(t) => return Ok(t);
+                    LoadTargetError::BuiltinTargetNotFound(_) => (),
+                    LoadTargetError::Other(e) => return Err(e),
```
Comment (oli-obk) [inline src/librustc_target/spec/mod.rs]: I'm not so sure about these load errors, it seems to me that they are a compiler bug. Either we panic on them and test them, or we allow the user to work around them by specifying a json file. I'd prefer the former.

Comment (gnzlbg) [reply src/librustc_target/spec/mod.rs]: > I'm not so sure about these load errors, it seems to me that they are a compiler bug.

When you target, say, an `ios` target, creating the Target executes code like this: https://github.com/rust-lang/rust/blob/master/src/librustc_target/spec/apple_ios_base.rs#L29

Which can fail, say on Linux, if that tool is not available. 

So in that case, load target failed, for a reason that AFAICT isn't a compiler bug, but an user bug: its running the compiler without the appropriate tools in the path.

Comment (gnzlbg) [reply src/librustc_target/spec/mod.rs]: I don't know if reporting those errors breaks somebody's workflow. Where they are relying, e.g., on the builtin target failing for external reasons, to override it with a json file. Honestly, that use case is better addressed with a `--target-file=file.json` (or just `--target=file.json`) CLI option that explicitly requests obtaining the target from a json file.

Comment (oli-obk) [reply src/librustc_target/spec/mod.rs]: that makes sense

**Pull request merged**